# Práctica 2: Patrones secuenciales

Imports necesarios para la realización de la práctica

In [13]:
import numpy as np
import pandas as pd
import sklearn as sk
import mlxtend as mlx
import apyori as apy
import matplotlib as mpl
from gsppy.gsp import GSP
from tabulate import tabulate

In [14]:
itemsets = pd.read_csv('splice.data', header=None)
#change the column names
itemsets.columns = ['empalme', 'donante', 'secuencia']


Pregunta Nº1:  Para la columna que contiene las secuencias de genes, generar una nueva columna (con la que trabajaremos posteriormente). Esta nueva columna debe contener los primeros dígitos de los genes siendo este número aleatorio contenido en el intervalo [4-10], es decir, cortaremos las secuencias para que su longitud sea aleatoria entre 4 y 10 dígitos. (1 punto)

In [15]:
#create a list for a new column
digitos = []
#remove the spaces in the sequence
itemsets['secuencia'] = itemsets['secuencia'].str.replace(' ', '')

for i in itemsets['secuencia']:

    #generate a random number between 4 and 10
    n = np.random.randint(4, 10)
    #get the first characters of the string
    digitos.append(i[:n])

itemsets['digitos'] = digitos

itemsets

empalme                  donante  \
0         EI         ATRINS-DONOR-521   
1         EI         ATRINS-DONOR-905   
2         EI         BABAPOE-DONOR-30   
3         EI        BABAPOE-DONOR-867   
4         EI       BABAPOE-DONOR-2817   
...      ...                      ...   
3185       N       ORAHBPSBD-NEG-2881   
3186       N        ORAINVOL-NEG-2161   
3187       N          ORARGIT-NEG-241   
3188       N           TARHBB-NEG-541   
3189       N          TARHBD-NEG-1981   

                                              secuencia    digitos  
0     CCAGCTGCATCACAGGAGGCCAGCGAGCAGGTCTGTTCCAAGGGCC...   CCAGCTGC  
1     AGACCCGCCGGGAGGCGGAGGACCTGCAGGGTGAGCCCCACCGCCC...      AGACC  
2     GAGGTGAAGGACGTCCTTCCCCAGGAGCCGGTGAGAAGCGCAGTCG...       GAGG  
3     GGGCTGCGTTGCTGGTCACATTCCTGGCAGGTATGGGGCGGGGCTT...     GGGCTG  
4     GCTCAGCCCCCAGGTCACCCAGGAACTGACGTGAGTGTCCCCATCC...       GCTC  
...                                                 ...        ...  
3185  TCTCTTCCCTTCCCCTCTCTCTTTCTTTCTTTTCTCTCCTCTTCTC...  TCTCTTCCC  
3186  GAGCTCCCAGAGCAGCAAGAGGGCCAGCTGAAGCACCTGGAGAAGC...   GAGCTCCC  
3187  TCTCGGGGGCGGCCGGCGCGGCGGGGAGCGGTCCCCGGCCGCGGCC...   TCTCGGGG  
3188  ATTCTACTTAGTAAACATAATTTCTTGTGCTAGATAACCAAATTAA...     ATTCTA  
3189  AGGCTGCCTATCAGAAGGTGGTGGCTGGTGTGGCTGCTGCTCTGGC...      AGGCT  

[3190 rows x 4 columns]

Pregunta Nº2: Comprueba que los datos son correctos: no hay clases vacías ni contenido repetido para la misma clase. (2 puntos)


In [16]:
itemsets.drop_duplicates().dropna()
itemsets = itemsets.iloc[:, 3]
itemsets

0        CCAGCTGC
1           AGACC
2            GAGG
3          GGGCTG
4            GCTC
          ...    
3185    TCTCTTCCC
3186     GAGCTCCC
3187     TCTCGGGG
3188       ATTCTA
3189        AGGCT
Name: digitos, Length: 3190, dtype: object

Pregunta Nº3: Prueba al menos dos configuraciones de soporte diferentes. (1 punto)

In [17]:
sequences = []
for index, row in itemsets.items():
    sequence = list(row)
    sequences.append(sequence)
print(sequences)
gspw = GSP(sequences)
gspw.maxPatternLength = 10  # establece la longitud máxima del patrón

gspw.search(0.1)

frequent_patterns_2 = gspw.freq_patterns.copy()

frequent_patterns_2


[['C', 'C', 'A', 'G', 'C', 'T', 'G', 'C'], ['A', 'G', 'A', 'C', 'C'], ['G', 'A', 'G', 'G'], ['G', 'G', 'G', 'C', 'T', 'G'], ['G', 'C', 'T', 'C'], ['C', 'A', 'G', 'A', 'C', 'T', 'G', 'G'], ['C', 'C', 'T', 'T', 'T', 'G', 'A', 'G'], ['C', 'C', 'C', 'T', 'C', 'G', 'T', 'G'], ['T', 'G', 'G', 'C', 'G', 'A', 'C', 'T'], ['A', 'A', 'G', 'C', 'T', 'G', 'A'], ['T', 'G', 'G', 'C', 'G', 'A'], ['A', 'A', 'G', 'C'], ['G', 'G', 'A', 'A', 'G', 'A'], ['A', 'A', 'G', 'C'], ['G', 'G', 'A', 'A', 'G', 'A', 'T'], ['A', 'A', 'G', 'C', 'T', 'G'], ['G', 'G', 'C', 'A', 'C'], ['T', 'T', 'G', 'C', 'T', 'C', 'T', 'G'], ['C', 'A', 'C', 'C', 'A', 'A'], ['A', 'C', 'A', 'G', 'A', 'G', 'G', 'A'], ['G', 'T', 'G', 'C', 'C', 'C', 'A'], ['C', 'A', 'C', 'G', 'A'], ['A', 'G', 'C', 'G', 'G', 'G', 'A', 'G', 'A'], ['A', 'T', 'G', 'A', 'G', 'A', 'A', 'G'], ['T', 'C', 'A', 'G'], ['G', 'A', 'T', 'C', 'C', 'G', 'C', 'C'], ['C', 'C', 'C', 'T', 'C', 'C', 'A', 'T', 'C'], ['C', 'C', 'C', 'A', 'A', 'G'], ['C', 'T', 'G', 'A', 'G', 'G', 'C

[{('C',): 2624, ('A',): 2492, ('G',): 2536, ('T',): 2512},
 {('A', 'A'): 778,
  ('C', 'T'): 1325,
  ('A', 'C'): 843,
  ('A', 'G'): 1052,
  ('C', 'G'): 453,
  ('A', 'T'): 819,
  ('C', 'C'): 1028,
  ('C', 'A'): 1099,
  ('G', 'C'): 1074,
  ('G', 'G'): 944,
  ('G', 'A'): 1015,
  ('G', 'T'): 666,
  ('T', 'A'): 569,
  ('T', 'C'): 1000,
  ('T', 'G'): 1188,
  ('T', 'T'): 721},
 {('C', 'A', 'G'): 333,
  ('G', 'C', 'T'): 353,
  ('G', 'G', 'C'): 319,
  ('C', 'C', 'T'): 381,
  ('C', 'T', 'C'): 355,
  ('T', 'G', 'G'): 351,
  ('C', 'T', 'G'): 449},
 {}]

In [19]:
gspw = GSP(sequences)
gspw.maxPatternLength = 10

gspw.search(0.3)

frequent_patterns_3 = gspw.freq_patterns.copy()

frequent_patterns_3

[{('G',): 2536, ('A',): 2492, ('T',): 2512, ('C',): 2624},
 {('G', 'A'): 1015,
  ('T', 'G'): 1188,
  ('A', 'G'): 1052,
  ('T', 'C'): 1000,
  ('G', 'C'): 1074,
  ('C', 'T'): 1325,
  ('C', 'C'): 1028,
  ('C', 'A'): 1099},
 {}]

Pregunta Nº4: Para una de ellas, interpreta algunos de los patrones secuenciales que te resulten curiosos. (1 punto)

- En la secuencia con un soporte de 0.1, se observan algunas secuencias interesantes, por ejemplo:

- La secuencia 'C' aparece con una frecuencia de 2624, lo que significa que es el nucleótido más común en la muestra.

- Las secuencias 'CAG' y 'CTG' aparecen con una frecuencia de 333 y 449, respectivamente, lo que indica que podrían estar relacionadas con algún evento biológico o proceso.

- Las secuencias 'AA', 'CC', 'GG' y 'TT' aparecen con frecuencias superiores a 700, lo que indica que podrían estar relacionadas con la formación de regiones altamente conservadas en el genoma.

- En la secuencia con un soporte de 0.3, se observan menos secuencias, lo que indica que se está utilizando un umbral de soporte más alto. Algunas secuencias interesantes incluyen:

- Las secuencias 'GC' y 'CA' aparecen con frecuencias superiores a 1000, lo que sugiere que pueden estar involucradas en algún tipo de interacción biológica.

- La secuencia 'CC' aparece con una frecuencia de 1028, lo que indica que podría estar involucrada en la formación de regiones altamente conservadas en el genoma.

En general, estas secuencias pueden ser de interés para los investigadores que estudian la genómica y la biología molecular.